src/bootchooser.c

```diff
@@ -1375,6 +1375,206 @@ static gboolean efi_get_state(RaucSlot* slot, gboolean *good, GError **error)
+	if (ret != 0) {
+		g_set_error(
+				error,
+				G_SPAWN_EXIT_ERROR,
+				ret,
+				"%s failed with wrong exit code", script_name);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* Wrapper for set commands accessing custom script
+ *
+ * @param cmd What to input as command to the custom backend. Mandatory.
+   Valid values:
+   - set-primary
+   - set-state
+ * @param bootname slot.bootname
+ * @param arg extra arguments if needed
+ * @param error Return location for a GError
+ */
+static gboolean custom_backend_set(const gchar *cmd, const gchar *bootname, const gchar *arg, GError **error)
+{
+	g_autoptr(GSubprocess) sub = NULL;
+	GError *ierror = NULL;
+	g_autoptr(GBytes) stdout_buf = NULL;
+	gchar *script_name = r_context()->config->custom_script;
+
+	g_return_val_if_fail(cmd, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (arg)
+		sub = g_subprocess_new(G_SUBPROCESS_FLAGS_NONE, &ierror, script_name, cmd, bootname, arg, NULL);
+	else
+		sub = g_subprocess_new(G_SUBPROCESS_FLAGS_NONE, &ierror, script_name, cmd, bootname, NULL);
+
+	if (!sub) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to start : %s", script_name);
+		return FALSE;
+	}
+
+	if (!g_subprocess_wait_check(sub, NULL, &ierror)) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to run: %s", script_name);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* Set slot status values */
+static gboolean custom_set_state(RaucSlot *slot, gboolean good, GError **error)
+{
+	GError *ierror = NULL;
+	const gchar *cmd = "set-state";
+	const gchar *state = good ? "good" : "bad";
+
+	g_return_val_if_fail(slot, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (!custom_backend_set(cmd, slot->bootname, state, &ierror) ) {
+		g_propagate_error(error, ierror);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* Get slot marked as primary one */
+static RaucSlot* custom_get_primary(GError **error)
+{
+	RaucSlot *slot;
+	GHashTableIter iter;
+	const gchar *cmd = "get-primary";
+	RaucSlot *primary = NULL;
+	GError *ierror = NULL;
+	g_autoptr(GString) ret_str = NULL;
+
+	if (!custom_backend_get(cmd, NULL, &ret_str, &ierror) ) {
+		g_propagate_error(error, ierror);
+		return NULL;
+	}
+
+	// Check result. Ensure returned bootname is in the list
+	g_hash_table_iter_init(&iter, r_context()->config->slots);
+	while (g_hash_table_iter_next(&iter, NULL, (gpointer*) &slot)) {
+		if (!slot->bootname)
+			continue;
+
+		if (g_strcmp0(ret_str->str, slot->bootname) == 0) {
+			primary = slot;
+		}
+	}
+
+	if (!primary) {
+		g_set_error_literal(
+				error,
+				R_BOOTCHOOSER_ERROR,
+				R_BOOTCHOOSER_ERROR_PARSE_FAILED,
+				"Unable to obtain primary element");
+	}
+
+	return primary;
+}
+
+/* Get state of current slot */
+static gboolean custom_get_state(RaucSlot *slot, gboolean *good, GError **error)
+{
+	GError *ierror = NULL;
+	const gchar *cmd = "get-state";
+	g_autoptr(GString) ret_str = NULL;
+	*good = FALSE;
```

Comment:
Hm, I don't think you're supposed to write to `*good` unless the function is succesful/returns true, at least judging from the other get_state implementations.

Reply:
I thought it made the most sense. But true; The other get_state does not write to it, unless they get proper feedback.

Removing the initial *good setting. Thanks